File: app/src/main/java/com/antyzero/smoksmog/logger/LoggerModule.java
package com.antyzero.smoksmog.logger;

import com.antyzero.smoksmog.BuildConfig;
import com.antyzero.smoksmog.user.User;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import smoksmog.logger.AndroidLogger;
import smoksmog.logger.Logger;

import static com.antyzero.smoksmog.logger.CrashlyticsLogger.ExceptionLevel.ERROR;

@Singleton
@Module
public class LoggerModule {

    @Provides
    @Singleton
    public Logger provideLogger(CrashlyticsLogger.ConfigurationCallback callback) {
        return BuildConfig.DEBUG ? new AndroidLogger() : new CrashlyticsLogger(ERROR, callback);
    }

    @Provides
    @Singleton
    public CrashlyticsLogger.ConfigurationCallback provideConfigurationCallback(User user) {
        return instance -> instance.setUserIdentifier(user.getIdentifier());
    }
}

File: app/src/main/kotlin/com/antyzero/smoksmog/logger/LoggerModule.kt
package com.antyzero.smoksmog.logger

import com.antyzero.smoksmog.BuildConfig
import com.antyzero.smoksmog.logger.CrashlyticsLogger.ExceptionLevel.ERROR
import com.antyzero.smoksmog.user.User
import com.crashlytics.android.core.CrashlyticsCore
import dagger.Module
import dagger.Provides
import smoksmog.logger.AndroidLogger
import smoksmog.logger.Logger
import javax.inject.Singleton

@Singleton
@Module
class LoggerModule {

    @Provides
    @Singleton
    internal fun provideLogger(callback: CrashlyticsLogger.ConfigurationCallback): Logger {
        return if (BuildConfig.DEBUG) AndroidLogger() else CrashlyticsLogger(ERROR, callback)
    }

    @Provides
    @Singleton
    internal fun provideConfigurationCallback(user: User): CrashlyticsLogger.ConfigurationCallback {
        return object : CrashlyticsLogger.ConfigurationCallback {
            override fun onConfiguration(instance: CrashlyticsCore) {
                instance.setUserIdentifier(user.identifier)
            }
        }
    }
}
